src/internal_events/buffer.rs

```diff
@@ -0,0 +1,18 @@
+use crate::internal_events::InternalEvent;
+use metrics::counter;
+
+#[derive(Debug)]
+pub struct BufferEventDropped;
+
+impl InternalEvent for BufferEventDropped {
```

Comment:
Based on our current structure, we have most events per-component. So if you wanted to track say, `GcpBufferEventDropped` that would be its own event. (https://github.com/timberio/vector/blob/5111f3a5b522fa06ca2ddf321e668246a78dc9fa/src/internal_events/elasticsearch.rs#L15)

This means quite some duplication though, so it would  be nice to either find another way, or you could create a small `macro_rules` and use that to do this: https://doc.rust-lang.org/reference/macros-by-example.html

Reply:
You may also be able to add some metadata to structs like `DiskBuffer` that contain the component name (`Option<&'static str>`) and have a builder `.with_component_name()` for opt in settings?